packages/core/src/behavior.ts

```diff
@@ -0,0 +1,360 @@
+import {
+  EventObject,
+  InvokeCallback,
+  Subscribable,
+  Unsubscribable,
+  InterpreterOptions,
+  Spawnable,
+  Observer
+} from './types';
+import { ActorRef } from './Actor';
+import {
+  toSCXMLEvent,
+  isPromiseLike,
+  isObservable,
+  isMachineNode,
+  isSCXMLEvent
+} from './utils';
+import { doneInvoke, error, actionTypes } from './actions';
+import { isFunction } from 'util';
+import { MachineNode } from './MachineNode';
+import { interpret, Interpreter } from './interpreter';
+import { State } from './State';
+
+export interface ActorContext {
+  self: ActorRef<any, any>; // TODO: use type params
+  name: string;
+}
+
+export const startSignal = Symbol('xstate.start');
+export const stopSignal = Symbol('xstate.stop');
+
+export type LifecycleSignal = typeof startSignal | typeof stopSignal;
+
+/**
+ * An object that expresses the behavior of an actor in reaction to received events,
+ * as well as an optionally emitted stream of values.
+ *
+ * @template TReceived The received event
+ * @template TEmitted The emitted value
+ */
+export interface Behavior<TReceived extends EventObject, TEmitted> {
+  receive: (
+    actorContext: ActorContext,
+    event: TReceived
+  ) => Behavior<TReceived, TEmitted>;
+  receiveSignal: (
+    actorContext: ActorContext,
+    signal: LifecycleSignal
+  ) => Behavior<TReceived, TEmitted>;
+  /**
+   * The initial emitted value
+   */
+  initial: TEmitted;
+  subscribe?: (observer: Observer<TEmitted>) => Unsubscribable | undefined;
+}
+
+export function createCallbackBehavior<TEvent extends EventObject>(
+  callback: InvokeCallback,
+  parent?: ActorRef<any>
+): Behavior<TEvent, undefined> {
+  let canceled = false;
+  const receivers = new Set<(e: EventObject) => void>();
+  let dispose;
+
+  const behavior: Behavior<TEvent, undefined> = {
+    receive: (_, event) => {
+      const plainEvent = isSCXMLEvent(event) ? event.data : event;
+      receivers.forEach((receiver) => receiver(plainEvent));
+
+      return behavior;
+    },
+    receiveSignal: (actorContext, signal) => {
+      if (signal === startSignal) {
+        dispose = callback(
+          (e) => {
+            if (canceled) {
+              return;
+            }
+
+            parent?.send(toSCXMLEvent(e, { origin: actorContext.self }));
+          },
+          (newListener) => {
+            receivers.add(newListener);
+          }
+        );
+
+        if (isPromiseLike(dispose)) {
+          dispose.then(
+            (resolved) => {
+              parent?.send(
+                toSCXMLEvent(doneInvoke(actorContext.name, resolved) as any, {
+                  origin: actorContext.self
+                })
+              );
+              canceled = true;
+            },
+            (errorData) => {
+              const errorEvent = error(actorContext.name, errorData);
+              parent?.send(
+                toSCXMLEvent(errorEvent, { origin: actorContext.self })
+              );
+              // TODO: handle error
```

Comment:
what kind of error is possible here? you already notify the "caller" actor through an event, i don't think there is particularly much else we could do here